lib/solaar/ui/pair_window.py

```diff
@@ -183,127 +144,107 @@ def _finish(assistant, receiver):
+def _show_passcode(assistant, receiver, passkey):
     if logger.isEnabledFor(logging.DEBUG):
-        logger.debug("%s fail: %s", receiver, error)
+        logger.debug("%s show passkey: %s", receiver, passkey)
+    name = receiver.pairing.device_name
+    authentication = receiver.pairing.device_authentication
+    intro_text = _("%(receiver_name)s: pair new device") % {"receiver_name": receiver.name}
+    page_text = _("Enter passcode on %(name)s.") % {"name": name}
+    page_text += "\n"
+    if authentication & 0x01:
+        page_text += _("Type %(passcode)s and then press the enter key.") % {"passcode": receiver.pairing.device_passkey}
+    else:
+        passcode = ", ".join(
+            [_("right") if bit == "1" else _("left") for bit in f"{int(receiver.pairing.device_passkey):010b}"]
+        )
+        page_text += _("Press %(code)s\nand then press left and right buttons simultaneously.") % {"code": passcode}
+    page = _create_page(assistant, Gtk.AssistantPageType.PROGRESS, intro_text, "preferences-desktop-peripherals", page_text)
+    assistant.set_page_complete(page, True)
+    assistant.next_page()
 
-    assistant.commit()
 
-    header = _("Pairing failed") + ": " + _(str(error)) + "."
-    if "timeout" in str(error):
-        text = _("Make sure your device is within range, and has a decent battery charge.")
-    elif str(error) == "device not supported":
-        text = _("A new device was detected, but it is not compatible with this receiver.")
-    elif "many" in str(error):
-        text = _("More paired devices than receiver can support.")
+def _create_assistant(receiver, ok, finish, title, text):
+    assistant = Gtk.Assistant()
+    assistant.set_title(title)
+    assistant.set_icon_name("list-add")
+    assistant.set_size_request(400, 240)
+    assistant.set_resizable(False)
+    assistant.set_role("pair-device")
+    if ok:
+        page_intro = _create_page(assistant, Gtk.AssistantPageType.PROGRESS, title, "preferences-desktop-peripherals", text)
+        spinner = Gtk.Spinner()
+        spinner.set_visible(True)
+        spinner.start()
+        page_intro.pack_end(spinner, True, True, 24)
+        assistant.set_page_complete(page_intro, True)
     else:
-        text = _("No further details are available about the error.")
-    _create_page(assistant, Gtk.AssistantPageType.SUMMARY, header, "dialog-error", text)
-
-    assistant.next_page()
-    assistant.commit()
+        page_intro = _create_failure_page(assistant, receiver.pairing.error)
+    assistant.connect("cancel", finish, receiver)
+    assistant.connect("close", finish, receiver)
+    return assistant
 
 
-def _pairing_succeeded(assistant, receiver, device):
-    assert device
-    if logger.isEnabledFor(logging.DEBUG):
-        logger.debug("%s success: %s", receiver, device)
+def _create_success_page(assistant, device):
+    def _check_encrypted(device, assistant, hbox):
+        if assistant.is_drawable() and device.link_encrypted is False:
```

Comment:
Possible to determine this, before the create_succes_page is called and just pass it as flag?

Reply:
No.  The device was just paired and might not be responding so there needs to be a delay here.   This has actually happened.

There is some other code - not a lot, but some - that does this sort of deferral, almost all added as the result of experience, often with low-end devices.